staging/src/k8s.io/client-go/discovery/cached/disk/size_limited_disk_cache.go

```diff
@@ -0,0 +1,208 @@
+	"k8s.io/klog/v2"
+)
+
+// SizeLimitedDiskCache is an implementation of httpcache.Cache that supplements
+// the in-memory map with persistent storage up to a pre-defined maxSize. After
+// the disk cache size breaches the size limit, the oldest files (by file
+// modification times) are evicted until the size limit is satisfied. This cache
+// is meant to be a substitute for "github.com/gregjones/httpcache/diskcache".
+type SizeLimitedDiskCache struct {
+	maxSize     uint64
+	currentSize uint64
+	fileInfos   []fileInfo
+	diskv       *diskv.Diskv
+}
+
+// Get returns the response corresponding to key if present.
+func (c *SizeLimitedDiskCache) Get(key string) (resp []byte, ok bool) {
+	key = keyToFilename(key)
+	resp, err := c.diskv.Read(key)
+	if err != nil {
+		return []byte{}, false
+	}
+	return resp, true
+}
+
+// Set saves a response to the cache as key, and updates the stored
+// fileInfos and currentSize. This function can cause cache evictions.
+func (c *SizeLimitedDiskCache) Set(key string, resp []byte) {
+	key = keyToFilename(key)
+	size := uint64(len(resp))
+	c.trimDiskCache(size)
+	err := c.diskv.WriteStream(key, bytes.NewReader(resp), true)
+	if err == nil {
+		c.appendFileInfo(fileInfo{name: key, size: size, modTime: time.Now()})
+	} else {
+		klog.Errorf("Error disk cache entry (%s):  [%v]", key, err)
+	}
+}
+
+// Delete removes the response with key from the cache, updating the
+// caches fileInfos and the caches currentSize.
+func (c *SizeLimitedDiskCache) Delete(key string) {
+	key = keyToFilename(key)
+	c.diskv.Erase(key)
+	c.removeFileInfo(key)
+}
+
+// Passing the size of a file that is going to be added, evicts entries
+// from the cache until the current cache size plus the passed size is below
+// the maximum cache disk size. Can update the cache fileInfos and currentSize
+// if evictions occur.
+func (c *SizeLimitedDiskCache) trimDiskCache(size uint64) {
+	for ((c.currentSize + size) > c.maxSize) && (len(c.fileInfos) > 0) {
+		fi := c.fileInfos[0] // Evict oldest file
+		err := c.diskv.Erase(fi.name)
+		if err == nil {
+			c.currentSize -= fi.size
+			c.fileInfos = c.fileInfos[1:]
+		} else {
+			// If disk cache eviction fails, just exit.
+			klog.Errorf("Error trimming disk cache: [%v]\n", err)
+			break
+		}
+	}
+}
+
+// IMPORTANT: This function MUST be the same as
+// "github.com/gregjones/httpcache/diskcache.keyToFilename()" in
+// order to maintain compatibility with existing http disk caches.
+func keyToFilename(key string) string {
+	h := md5.New()
+	io.WriteString(h, key)
+	return hex.EncodeToString(h.Sum(nil))
+}
+
+// NewSizeLimitedDiskCache returns a new Cache up to "maxSize" number
+// of total file bytes in the cache. Uses the provided Diskv as
+// underlying storage. Walks the entire tree of files in the cache.
+// Can cause evictions from the cache if the initial size of the cache
+// is beyond "maxSize".
+func NewSizeLimitedDiskCache(maxSize uint64, diskv *diskv.Diskv) *SizeLimitedDiskCache {
```

Comment:
Since this abstraction doesn't work for all settings of diskv, let's make sure that this owns the creation of diskv.

Reply:
Instead, I planned on doing verification checks (e.g. check the `Transform` field is not set) on `Diskv` in this initialization function.

Reply:
After further discussion offline, we agreed to pull the `Diskv` creation into `NewSizeLimitedDiskCache()` instead of passing it in. We will do this to ensure the `Diskv` can not be created with incompatible parameters.